doc/tagsearch.1
.\" Automatically generated by Pandoc 2.11.3
.\"
.TH "tagsearch" "1" "2021-03-21" "Search and report on tags" ""
.hy
.SH NAME
.PP
tagsearch
.SH SYNOPSIS
\fBtagsearch\fR [\fB\--root\fR \fIroot\fR] \fICOMMAND\fR [\fICOMMAND-OPTIONS\fR] [\fICOMMAND-FLAGS\fR]
.PP
.RS 4
\fIroot\fR is a different root directory. Otherwise, \fBtagsearch\fR will 
operate on all files under the current directory.
.RE

.SH COMMANDS
\fIfile-tags\fR [\fB\-c|--count\fR] [\fIfiles...\fR]
.RS 4
Show the tags from each file. \fI\-\-count\fR will show how many times each file is used across these files
.RE
.PP
\fIfiles\fR [\fB\-o|\-\-or\fR] [\fB\-\-vim\fR] [\fB\-\-not\fR \fInot...\fR] [\fItags...\fR]
.RS 4
\fB\-o\fR will match \fIANY\fR, rather than \fIALL\fR, tags
.P
\fB\-\-vim\fR will output a format suitable for use with \fBvimgrep\fR.
.P
\fB\-\-not\fR \fInot...\fR will exclude results which have '\fInot\fR' tags
.P
\fItags...\fR is the tags you want to find files for.
.RE
.PP
\fIhelp\fR
.RS 4
Display a brief help string listing available commands
.RE

.PP
\fIsimilar-tags\fR
.RS 4
Show tags that are potential near-duplicates (mainly searches for differences in pluralisation)
.RE

.PP
\fItags\fR [\fIOPTIONS\fR] [\fIFLAGS\fR] [\fB\-\-not\fR \fInot...\fR] [\fITAGS...\fR]
.RS 4
Show all tags from files with tags that match filter words
.RE
.PP
\fIuntagged\fR
.RS 4
description
.RE

.SH DESCRIPTION
Search for all \f[B]keyword\f[R] tags for files under the current directory.

.SH FLAGS
.TP
\f[B]-c\f[R], \f[B]--count\f[R]
Show count of tags
.TP
\f[B]-h\f[R], \f[B]--help\f[R]
Prints help information
.TP
\f[B]-l\f[R], \f[B]--list\f[R]
List all tags for files matching keywords
.TP
\f[B]--long\f[R]
Long list (e.g.\ tall) all tags for files matching keywords
.TP
\f[B]-o\f[R], \f[B]--or-filter\f[R]
Filter using ANY, rather than ALL keywords
.TP
\f[B]--similar-tags\f[R]
Show similar tags
.TP
\f[B]-u\f[R], \f[B]--untagged\f[R]
Show untagged files
.TP
\f[B]-V\f[R], \f[B]--version\f[R]
Prints version information
.TP
\f[B]-v\f[R], \f[B]--vim\f[R]
Output format suitable for vim quickfix
.TP
\f[B]--not NOT\&...\f[R]
Keywords to inverse filter (i.e.\ ignore matching files)
.SH NOTES
.PP
A \f[B]tag\f[R] is something that matches the regexp
\f[B]\[at][a-zA-Z1-9]+\f[R], i.e.\ \f[B]\[at]\f[R] followed by anything
\f[B]alphanumeric\f[R].
.IP
.nf
\f[C]
\[at]rust
\[at]linux
\[at]c99
\f[R]
.fi

.SH EXAMPLES

.IP 1
Find all files in the current directory tagged \fB@stoicism\fR \fIand\fR \fB@philosophy\fR
.nf
\fBtagsearch files\fR \fIstoicism\fR \fIphilosophy\fR
.fi
.PP

.IP 2 
Find all files in the current directory tagged \fIeither\fR \fB@rust\fR or \fB@golang\fR
.nf
\fBtagsearch files\fR \fI\-\-or\fR \fIrust\fR \fIgolang\fR
.fi
.PP


.IP 3
List all tags in the current directory
.nf
\fBtagsearch tags\fR
.fi
.PP

.IP 4
List all tags in a different directory
.nf
\fBtagsearch --root\fR \fI~/somewhere/ tags\fR
.fi




.SH AUTHORS
.PP
Chris Davison <c.jr.davison@gmail.com>


tagsearch.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"sort"
	"strings"

	"github.com/bmatcuk/doublestar"
	"gopkg.in/alecthomas/kingpin.v2"
)

const VERSION = "0.3.0"

var (
	list      = kingpin.Flag("list", "List tags").Short('l').Bool()
	verbose   = kingpin.Flag("verbose", "List tags with count of files").Short('v').Bool()
	summarise = kingpin.Flag("summarise", "List tags and matching files").Short('s').Bool()
	files     = kingpin.Arg("files", "Files to search/summarise").Strings()
	keywords  = kingpin.Flag("keywords", "Keywords to filter for").Short('k').Strings()
	andFilter = kingpin.Flag("and-filter", "Filter using ALL tags (or default ANY match)").Bool()
	version   = kingpin.Flag("version", "Show version").Bool()
)

func getAllTags(files []string) taglistWithCount {
	tagsSeen := make(map[string]int)
	for _, filename := range files {
		for _, tag := range getTagsForFile(filename) {
			tagsSeen[tag] += 1
		}
	}
	taglist := make(taglistWithCount, 0)
	for tag, count := range tagsSeen {
		taglist = append(taglist, tagWithCount{tag, count})
	}
	sort.Sort(sort.Reverse(taglist))
	return taglist
}

func listTags(files []string, verbose bool) {
	if verbose {
		for _, countedTag := range getAllTags(files) {
			fmt.Printf("%5d %s\n", countedTag.count, countedTag.tag)
		}

	} else {
		tagnames := make([]string, 0)
		for _, countedTag := range getAllTags(files) {
			tagnames = append(tagnames, countedTag.tag)
		}
		fmt.Println(strings.Join(tagnames, ", "))
	}

}

func getTagsForFile(filename string) []string {
	contents, err := ioutil.ReadFile(filename)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return []string{}
	}
	rx := regexp.MustCompile(`(?:^|\s)@([a-zA-Z_0-9\-]+)`)
	matches := rx.FindAllStringSubmatch(string(contents), -1)
	matchesSeen := make(map[string]bool)
	for _, match := range matches {
		for _, keyword := range match[1:] {
			matchesSeen[keyword] = true
		}
	}
	onlyGroups := make([]string, 0)
	for kw, _ := range matchesSeen {
		onlyGroups = append(onlyGroups, strings.ToLower(kw))
	}
	return onlyGroups
}

func keywordMapOr(keywordToFile map[string][]string, keywords []string) []string {
	results := make([]string, 0)
	filenamesSeen := make(map[string]bool)
	for _, kw := range keywords {
		for _, filename := range keywordToFile[kw] {
			filenamesSeen[filename] = true
		}
	}
	for filename, _ := range filenamesSeen {
		results = append(results, filename)
	}
	return results
}

func keywordMapAnd(keywordToFile map[string][]string, keywords []string) []string {
	results := make([]string, 0)
	countOfKeywords := make(map[string]int)
	nilMatches := make([]string, 0)
	for _, kw := range keywords {
		files, ok := keywordToFile[kw]
		if !ok {
			fmt.Printf("No matches for `%s`\n", kw)
			nilMatches = append(nilMatches, kw)
			continue
		}
		for _, file := range files {
			countOfKeywords[file] += 1
		}
	}
	if nilMatches == nil {
		fmt.Println("No matches for: ", strings.Join(nilMatches, ", "))
	}
	for filename, count := range countOfKeywords {
		if count == len(keywords) {
			results = append(results, filename)
		}
	}
	return results
}

func main() {
	kingpin.Parse()
	if *version {
		fmt.Printf("tagsearch v%s\n", VERSION)
		os.Exit(0)
	}
	var err error
	if *files == nil {
		*files, err = doublestar.Glob("**/*txt")
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		filesMd, err := doublestar.Glob("**/*md")
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
		}
		*files = append(*files, filesMd...)
	}
	if *verbose || *list || (*keywords == nil && !*summarise) {
		listTags(*files, *verbose)
		os.Exit(0)
	}

	keywordToFile := make(map[string][]string, 0)
	for _, filename := range *files {
		for _, tag := range getTagsForFile(filename) {
			keywordToFile[tag] = append(keywordToFile[tag], filename)
		}
	}
	if *summarise {
		for kw, files := range keywordToFile {
			fmt.Println(kw)
			for _, filename := range files {
				fmt.Println("\t", filename)
			}
		}
		os.Exit(0)
	}
	matchingFiles := make([]string, 0)
	if *andFilter {
		matchingFiles = keywordMapAnd(keywordToFile, *keywords)
	} else {
		matchingFiles = keywordMapOr(keywordToFile, *keywords)
	}
	for _, filename := range matchingFiles {
		fmt.Println(filename)
	}
}
